### PR TITLE
Add cross-compile support via xgo

### PR DESCRIPTION
**Pull request by konez2k**
Added `-x <platform>` flag to build command, supported platforms are the same as xgo.
You need to install `techknowlogick/xgo` via `go get src.techknowlogick.com/xgo` (this repo supports go modules).
The platform packaging is also supported using the same xgo docker image, for windows windres tool.

By default the executables are stored inside the `build` folder, if you dont specify the architecture, the arch is automatically set to amd64.

Cross-compilation has been tested on Linux.
Packed builds tested on MacOS 10.14 and Windows 10 via:

`wails build -p -x windows/386`.
`wails build -p -x windows`.
`wails build -p -x darwin`.

### CODE REVIEW

**Comment by tmclane** [comment]
Holy cow this is awesome!

**Comment by tmclane** [comment]
I just tested building the `wailskiosk` example I put together on a Linux box for a Mac.
Works perfectly!
Good job!

**Comment by leaanthony** [comment]
This is *incredible*! Thank you so much for doing this. Will test and
hopefully merge tonight.

🚀🚀🚀

File: cmd/package.go [hunk 1]
```diff
@@ -192,16 +217,36 @@ func (b *PackageHelper) PackageWindows(po *ProjectOptions, cleanUp bool) error {
...
+		args := []string{
+			"docker", "run", "--rm",
+			"-v", folder + ":/build",
+			"--entrypoint", "/bin/sh",
+			"techknowlogick/xgo",
+			"-c", "/usr/bin/x86_64-w64-mingw32-windres -o /build/" + basename + "-res.syso /build/" + basename + ".rc",
+		}
+
+		if err := NewProgramHelper().RunCommandArray(args); err != nil {
+			return err
```
Comment (leaanthony) [inline cmd/package.go]: What might be nice is to report to the user that docker doesn't exist and to install it, but I'm guessing the default message would be something similar to that...

Comment (konez2k) [reply cmd/package.go]: same as if xgo is not installed, they should get a command not found error.
There is already a check for mewn, so adding the 2 checks (xgo/docker) shouldnt take much time, let me know if I've to add them in this PR or a new once (after this one is merged).

I was thinking also on using the `builds` folder as a standard directory where to put both cross-compiled and normal builds, less mess on project folder.

Comment (tmclane) [reply cmd/package.go]: I would recommend adding them to this PR.  I don't see any reason you shouldn't `fixup` the commit into the one on this branch to resolve this conversation.

File: cmd/helpers.go [hunk 2]
```diff
@@ -77,8 +102,23 @@ func BuildApplication(binaryName string, forceRebuild bool, buildMode string, pa
 	packSpinner.SetSpinSpeed(50)
 	packSpinner.Start()
 
+	// embed resources
+	targetFiles, err := EmbedAssets()
+	if err != nil {
+		return err
```
Comment (tmclane) [inline cmd/helpers.go]: Does this trigger logging somewhere higher up if it fails?

**Comment by tmclane** [comment]
One thing that did happen is the following when cross-compiling for Darwin:

```
$ wails build -p -x darwin
Wails v1.0.2 - Packaging Application

✓ Skipped frontend dependencies (-f to force rebuild)
✓ Building frontend...
✓ Ensuring Dependencies are up to date...
✓ Packing + Compiling project...
✗ Generating .app
Error: rename /home/tmclane/Projects/W/wailskiosk/build/wailskiosk-darwin-10.6-amd64 /home/tmclane/Projects/W/wailskiosk/build/wailskiosk: permission denied
```

**Comment by tmclane** [comment]
Ah!  

The output files are owned by root:

```
tmclane@xps15-9560:~/Projects/W/wailskiosk$ ls -lisa build/
total 23540
79705003     4 drwxr-xr-x 2 root    root        4096 Feb 20 08:13 .
77207270     4 drwx------ 5 tmclane tmclane     4096 Feb 20 08:13 ..
79705060 11972 -rwxr-xr-x 1 root    root    12256196 Feb 20 08:13 wailskiosk-darwin-10.6-amd64
79705220 11560 -rwxr-xr-x 1 root    root    11837440 Feb 19 10:30 wailskiosk-windows-4.0-amd64.exe
```

I am sure a small tweak to the Docker container execution can fix this.  Had to do something similar for a work project.

**Comment by konez2k** [comment]
just retested, its due to docker and its solvable by creating the build directory before launching xgo, will add a commit to fix this 

> Ah!
> 
> The output files are owned by root:
> 
> ```
> tmclane@xps15-9560:~/Projects/W/wailskiosk$ ls -lisa build/
> total 23540
> 79705003     4 drwxr-xr-x 2 root    root        4096 Feb 20 08:13 .
> 77207270     4 drwx------ 5 tmclane tmclane     4096 Feb 20 08:13 ..
> 79705060 11972 -rwxr-xr-x 1 root    root    12256196 Feb 20 08:13 wailskiosk-darwin-10.6-amd64
> 79705220 11560 -rwxr-xr-x 1 root    root    11837440 Feb 19 10:30 wailskiosk-windows-4.0-amd64.exe
> ```
> 
> I am sure a small tweak to the Docker container execution can fix this. Had to do something similar for a work project.

File: cmd/helpers.go (hunk 2 above)
Comment (konez2k) [reply cmd/helpers.go]: yes, its sent back to the calling cmd command and displayed on console as a build error

**Comment by tmclane** [comment]
> just retested, its due to docker and its solvable by creating the build directory before launching xgo, will add a commit to fix this
> 
> > Ah!
> > The output files are owned by root:
> > ```
> > tmclane@xps15-9560:~/Projects/W/wailskiosk$ ls -lisa build/
> > total 23540
> > 79705003     4 drwxr-xr-x 2 root    root        4096 Feb 20 08:13 .
> > 77207270     4 drwx------ 5 tmclane tmclane     4096 Feb 20 08:13 ..
> > 79705060 11972 -rwxr-xr-x 1 root    root    12256196 Feb 20 08:13 wailskiosk-darwin-10.6-amd64
> > 79705220 11560 -rwxr-xr-x 1 root    root    11837440 Feb 19 10:30 wailskiosk-windows-4.0-amd64.exe
> > ```
> > 
> > 
> > I am sure a small tweak to the Docker container execution can fix this. Had to do something similar for a work project.

It does appear that creating the `build` directory beforehand locally allows the `.app` package to be createable.

**Comment by tmclane** [comment]
Just tested with the latest two commits!  Darwin App bundles work fine without any special handling beforehand now.

**Comment by leaanthony** [comment]
Hey there. I'm really excited about this feature and am keen to test it thoroughly. I've created this page here to keep track of the support/test matrix. If it's too hard to edit the wiki, we can do it here and I can update it. I'm planning on doing as much as I can on this at the weekend. I think it would be good to record the OS versions too.

**Comment by leaanthony** [comment]
Could you please make this PR against the `develop` branch. Thanks!

**Comment by leaanthony** [comment]
This is working really well. One bit of feedback is that a linux build will fail as it does not contain the webkit/gtk libraries. I'm considering whether we just publish a custom xgo build tailored for this?

**Comment by leaanthony** [comment]
Just noticed, windows/386 fails to pack:
```
/usr/local/go/pkg/tool/linux_amd64/link: running i686-w64-mingw32-gcc-posix failed: exit status 1
/tmp/go-link-309413631/000000.o: file not recognized: File format not recognized
collect2: error: ld returned 1 exit status
```

**Comment by leaanthony** [comment]
I've created a new branch [cross-compilation](https://github.com/wailsapp/wails/tree/cross-compilation) that has been merged with latest develop and a warning for unsupported platforms. You may want to update your PR to be against this branch then we can work on finalising it.

**Comment by konez2k** [comment]
```
# pkg-config --cflags  -- gtk+-3.0 webkit2gtk-4.0
Package gtk+-3.0 was not found in the pkg-config search path.
Perhaps you should add the directory containing `gtk+-3.0.pc'
to the PKG_CONFIG_PATH environment variable
No package 'gtk+-3.0' found
Package webkit2gtk-4.0 was not found in the pkg-config search path.
Perhaps you should add the directory containing `webkit2gtk-4.0.pc'
to the PKG_CONFIG_PATH environment variable
No package 'webkit2gtk-4.0' found
pkg-config: exit status 1
```

**Comment by konez2k** [comment]
Build the custom xgo image
```
docker build -t xgo Docker/
```
You can push this image to docker hub, so we can pull it later
(TODO: test if its pulled automatically via xgo or must be pulled on first run)

Temporary tests:
```
xgo -v -image xgo -targets=linux/amd64 ./
```
works, no html (not embedded) 

```
xgo -v -image xgo -targets=linux/386 ./
```

build fails, probably the webview dont set the 32bit CFLAGS (-m32) via CGO.
Is 32bit support still a thing under linux? ;)

**Comment by konez2k** [comment]
windows/386 works on my side, more infos on source platform / build / docker image

**Comment by leaanthony** [comment]
Just an FYI, I've been *super* busy last few weeks to look at this. I will absolutely look at this at the weekend and thanks for your patience :+1:

**Comment by leaanthony** [comment]
This is my next priority for V1. I just need some time for it.

**Comment by zhuqiyun** [comment]
how to compile for rasberry？ what‘s the exact cmd？could cross build on win10？

**Comment by oliverpool** [comment]
@zhuqiyun maybe something like `wails build -p -x linux/arm` (be sure to be on this branch and check the requirements of the first post).

List of supported archs: https://hub.docker.com/r/karalabe/xgo-latest

**Comment by zhuqiyun** [comment]
> @zhuqiyun maybe something like `wails build -p -x linux/arm` (be sure to be on this branch and check the requirements of the first post).
> 
> List of supported archs: https://hub.docker.com/r/karalabe/xgo-latest

thanks  ，i will try

**Comment by tmclane** [comment]
@zhuqiyun 
FYI: I am working on this now to get RaspberryPi support implemented.  It doesn't currently work.
I found a bug in the base image used by this process.

**Comment by tmclane** [comment]
Closing this since I included all your changes in my PR #377.

**Comment by zhuqiyun** [comment]
> @zhuqiyun
> FYI: I am working on this now to get RaspberryPi support implemented. It doesn't currently work.
> I found a bug in the base image used by this process.

which branch or when  raspberry cross could be used, thanks

**Comment by tmclane** [comment]
It was merged into `develop` and is included in version 1.5.0 which was just released.

**Pull request closed without merging**